sympy/functions/special/bsplines.py

```diff
@@ -202,3 +202,60 @@ def bspline_basis_set(d, knots, x):
     """
     n_splines = len(knots) - d - 1
     return [bspline_basis(d, knots, i, x) for i in range(n_splines)]
+
+def interpolating_spline(d, x, xv='x', yv='y'):
+    """Return B-splines at ``x`` of degree ``d`` for given ``xv`` and corresponding ``yv`` values.
+
+    This function returns a list of Piecewise polynomials that are the
+    B-splines of degree ``d`` for the given points. This function
+    calls ``bspline_basis_set(n, knots, x)`` for different values of ``n``.
+
+    Examples
+    ========
+
+    >>> from sympy interpolating_spline
+    >>> from sympy.abc import x
+    >>> d = 2
+    >>> splines=interpolating_spline(d, x, [1,2,3], [1,2,3])
+    >>> splines
+    [4*Piecewise((-x**2/4 + x - 3/4, (x >= 1) & (x <= 3)), (0, True)) 
+     + Piecewise((x**2/4 - 3*x/2 + 9/4, (x >= 1) & (x <= 3)), (0, True)) 
+     + 3*Piecewise((x**2/4 - x/2 + 1/4, (x >= 1) & (x <= 3)), (0, True))
+    ] 
+
+    See Also
+    ========
+
+    bsplines_basis_set
+    """
+    from sympy import symbols
+    from sympy.solvers.solveset import linsolve
+    from sympy.matrices.dense import Matrix
+    
+    
+    ## IF no xv and yv values are passed as a argument
+    if isinstance(xv, str):
+        xv = symbols("%s:%s" % (xv, d))
+
+    if isinstance(yv, str):
+        yv = symbols("%s:%s" % (yv, d))
+
+
+    ##EVALUATING knots value
+    if d % 2 == 1:
+        j = (d+1) // 2
+        interior_knots = xv[j:-j]
+    else:
+        j = d // 2
+        interior_knots = [Rational(a+b, 2) for a, b in zip(xv[j:-j-1], xv[j+1:-j])]
+
+    knots = [xv[0]] * (d+1) + list(interior_knots) + [xv[-1]] * (d+1)
+
+    basis = bspline_basis_set(d, knots, x)
+
+    A = [[b.subs(x, v) for b in basis] for v in xv]
+    coeff = linsolve((Matrix(A), Matrix(yv)), symbols('c0:{}'.format(len(basis))))
+    spline = sum([c*b for c, b in zip(list(coeff)[0], basis)])
```

Comment:
Try applying piecewise_fold to `spline` here, it should be able to combine the Piecewise objects together. See  http://docs.sympy.org/latest/modules/functions/elementary.html#sympy.functions.elementary.piecewise.piecewise_fold

Reply:
I tried it just now. It does combine the terms, but (a) takes a **very** long time, and (b) the output still has some issues. This is what it gives for the data in my SO post:
```
Piecewise((1168*x**3/405 - 397*x**2/27 + 3277*x/135 - 3161/405, Eq(x, 2)), (23*x**3/15 - 33*x**2/5 + 121*x/15 + 3, (x >= 0) & (x <= 2)), (-13*x**3/24 + 219*x**2/40 - 359*x/24 + 689/40, Eq(x, 3)), (-2*x**3/3 + 33*x**2/5 - 55*x/3 + 103/5, (x >= 2) & (x <= 3)), (-28*x**3/15 + 87*x**2/5 - 761*x/15 + 53, (x >= 3) & (x <= 5)), (0, True))
``` 
The special casing Eq(x, 2) and Eq(x, 3) should not be there at all.  

I think the right approach here is to fold the expression "manually" by looping through those Piecewise objects and adding the functions that correspond to the same condition. As a start, read about the [Piecewise class](http://docs.sympy.org/latest/modules/functions/elementary.html#piecewise).

Reply:
I tried the following, which seems to work. See if it helps to get a good form of the result. First, a cosmetic change to `linsolve`:  using % formatting (consistent with the rest of SymPy codebase) and argument `cls=Dummy` to make dummy symbols instead of regular symbols. 

Given a Piecewise object `b`, we can access the pieces with `b.args`, which is a tuple of tuples (expression, condition). This is used in two ways: 

1. Make a set of all `intervals` that appear in the splines
2. Turn each basis spline into a dictionary where the keys are intervals. These are in `basis_dicts`.

Then, instead of adding Piecewise objects themselves, we go over the `intervals`; for each interval, add the values of basis splines (with coefficients) for that interval. The is the sum called `piece`. The pieces are packaged in tuples with the corresponding intervals. The returned object is a single Piecewise made of these tuples.   

```
coeff = linsolve((Matrix(A), Matrix(yv)), symbols('c0:%d' % len(xv), cls=Dummy))
coeff = list(coeff)[0]
intervals = set([c for b in basis for (e, c) in b.args if c != True])
basis_dicts = [dict((c, e) for (e, c) in b.args) for b in basis]
spline = []
for i in intervals:
    piece = sum([c*d.get(i, S.Zero) for (c, d) in zip(coeff, basis_dicts)], S.Zero)
    spline.append((piece, i))
spline.append((S.Zero, True))
print(Piecewise(*spline))     # this is the spline to return
```

Reply:
Oh sorry for all these noob mistakes, my first pull request :)
I didn't get what is the problem with special casing Eq(x, 2) and Eq(x, 3). Isn't just same as x=2 and x=3 respectively?

Reply:
The problem is that they should not be there. A formula for spline should not have "this when x=2" or "that when x=3" clauses. It should have "this formula if 2 <= x <= 3" and "that formula when 3 <= x <= 5".

Besides, piecewise_fold does take incredibly long, much longer than the rest of the computation. So I think the manual folding, as sketched above, is the way to go.